pkg/cmd/cli/cmd/login.go

```diff
@@ -0,0 +1,248 @@
+				if requiresNewLogin {
+					clientCfg.BearerToken = ""
+					accessToken, err := tokencmd.RequestToken(clientCfg, os.Stdin, options.username, options.password)
+					checkErr(err)
+
+					if userFullName, err := whoami(clientCfg); err == nil {
+						err = config.UpdateConfigFile(userFullName, accessToken, f.OpenShiftClientConfig.KClientConfig, configStore)
+						checkErr(err)
+						fmt.Printf("Logged into %v as %v\n", clientCfg.Host, userFullName)
+					} else {
+						glog.Fatalf("%v\n", err)
+					}
+				}
+			}
+
+			// select a project to use
+			// if the user only has one project, use that one
+			// if user has multiple, take the first and print the others
+			// if --project were provided, use that one
+			oClient, err := client.New(clientCfg)
+			checkErr(err)
+			// TODO must properly handle policies (projects an user belongs to) https://github.com/openshift/origin/pull/971
+			projects, err := oClient.Projects().List(labels.Everything(), labels.Everything())
+			checkErr(err)
+
+			if size := len(projects.Items); size > 0 {
+				if options.projectProvided() {
+					project, err := oClient.Projects().Get(options.project)
+					checkErr(err)
+					currentCtx := configStore.Config.Contexts[configStore.Config.CurrentContext]
+
+					currentCtx.Namespace = project.Name
+					configStore.Config.Contexts[configStore.Config.CurrentContext] = currentCtx
+
+					if err = kclientcmd.WriteToFile(*configStore.Config, configStore.Path); err != nil {
+						glog.Fatalf("Error saving config to file: %v", err)
+					}
+				}
+				if size > 1 {
+					projectsNames := make([]string, size)
+					for i, project := range projects.Items {
+						projectsNames[i] = project.Name
+					}
+					fmt.Printf("Your projects are: %v. You can switch between them at any time using 'osc project <project-name>'.\n", strings.Join(projectsNames, ", ")) // TODO switch projects TODO parameterize cmd name
+				}
+				current, err := f.OpenShiftClientConfig.Namespace()
+				checkErr(err)
+				fmt.Printf("Now using project '%v'\n", current)
+			} else {
+				// TODO handle no projects
+			}
+		},
+	}
+
+	cmds.Flags().StringVarP(&options.username, "username", "u", "", "Username, will prompt if not provided")
+	cmds.Flags().StringVarP(&options.password, "password", "p", "", "Password, will prompt if not provided")
+	cmds.Flags().StringVar(&options.project, "project", "", "If provided the client will switch to use the provided project after logging in")
+	// TODO should *probably* explicitly expose the flags below as local (currently global)
+	//cmds.Flags().StringVar(&options.context, "context", "", "Use a specific config context to log in")
+	//cmds.Flags().StringVar(&options.server, "server", "", "The address and port of the API server to log in to")
+	return cmds
+}
+
+func whoami(clientCfg *kclient.Config) (string, error) {
+	oClient, err := client.New(clientCfg)
+	if err != nil {
+		return "", err
+	}
+
+	me, err := oClient.Users().Get("~")
```

Comment:
I don't think we should depend on this - not every valid authenticated account will have a user record.  For now, we should be using a more generic endpoint.